metadata-io/src/main/java/com/linkedin/metadata/changeprocessor/ChangeStreamProcessor.java

```diff
@@ -0,0 +1,107 @@
+package com.linkedin.metadata.changeprocessor;
+
+import com.linkedin.data.template.RecordTemplate;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+
+/**
+ * Responsible for routing events to the correct registered change processors. Processors are registered based on the
+ * entity/aspect they apply to. If none are supplied then the proposed change is returned.
+ */
+@Slf4j
+public class ChangeStreamProcessor {
+
+  private final Comparator<ChangeProcessor> _processorComparator = Comparator.comparing(ChangeProcessor::getPriority);
+  private final Map<String, SortedSet<ChangeProcessor>> _preProcessors = new ConcurrentHashMap<>();
+  private final Map<String, SortedSet<ChangeProcessor>> _postProcessors = new ConcurrentHashMap<>();
+
+  public void registerPreProcessor(String entityName, String aspectName, ChangeProcessor processor) {
+    registerProcessor(entityName, aspectName, processor, _preProcessors);
+  }
+
+  public void registerPostProcessor(String entityName, String aspectName, ChangeProcessor processor) {
+    registerProcessor(entityName, aspectName, processor, _postProcessors);
+  }
+
+  public ChangeResult preProcessChange(String entityName, String aspectName, RecordTemplate previousAspect,
+      RecordTemplate newAspect) {
+    return process(entityName, aspectName, previousAspect, newAspect, _preProcessors);
+  }
+
+  public ChangeResult postProcessChange(String entityName, String aspectName, RecordTemplate previousAspect,
+      RecordTemplate newAspect) {
+    return process(entityName, aspectName, previousAspect, newAspect, _postProcessors);
+  }
+
+  private void registerProcessor(String entityName, String aspectName, ChangeProcessor processor,
+      Map<String, SortedSet<ChangeProcessor>> processorMap) {
```

Comment:
You could emphasize here that `processorMap` is expected to be a `ConcurrentMap`. Ideally, the value type should also indicate it is thread-safe, but there is no interface for the `synchronizedSortedSet`, so this is the best you can do I think.